tests/util/helper.py

```diff
@@ -482,22 +482,43 @@ def get_preview_sql_for_series(series_obj, *args, **kwargs):
     return SampleMixin.preview_sql(series_obj, *args, **kwargs)
 
 
-def check_decomposed_graph_output_node_hash(
-    feature_model, offline_store_ingest_query_graph_extractor_output=None
-):
+def check_decomposed_graph_output_node_hash(feature_model, output=None):
     """Check the output node hash of the decomposed graph is the same as the original graph"""
-    if offline_store_ingest_query_graph_extractor_output is None:
-        extractor = OfflineStoreIngestQueryGraphExtractor(graph=feature_model.graph)
-        offline_store_ingest_query_graph_extractor_output = extractor.extract(
-            node=feature_model.node, relationships_info=feature_model.relationships_info
+    if output is None:
+        transformer = OfflineStoreIngestQueryGraphTransformer(graph=feature_model.graph)
+        output = transformer.transform(
+            target_node=feature_model.node,
+            entity_id_to_serving_name={},
+            relationships_info=feature_model.relationships_info,
+            feature_name=feature_model.name,
         )
 
+    if output.is_decomposed is False:
+        assert output.node_name_map == {}
+        return
+
+    # check all the original graph's node_name can be found in node_name_map
+    assert set(feature_model.graph.nodes_map.keys()) == set(output.node_name_map.keys())
+
+    # check that no unexpected node type is present in the decomposed graph
+    unexpected_node_types = {
+        NodeType.INPUT,
+        NodeType.GROUPBY,
+        NodeType.ITEM_GROUPBY,
+        NodeType.AGGREGATE_AS_AT,
+        NodeType.LOOKUP,
+        NodeType.LOOKUP_TARGET,
+        NodeType.JOIN,
+        NodeType.JOIN_FEATURE,
+    }
+    for node in output.graph.nodes:
+        assert node.type not in unexpected_node_types, "Unexpected node type in decomposed graph"
```

Comment:
Curious if this assertion would fail in existing test cases before the fix?

Reply:
Good catch, add tests in `tests/unit/api/test_offline_ingest_query_graph.py` to cover the fail cases.